cocotb/regression.py

```diff
@@ -130,111 +117,135 @@ def initialise(self):
             self._cov = coverage.coverage(branch=True, omit=["*cocotb*"])
             self._cov.start()
 
-        # Setup DUT object
-        #######################
+        # Test Discovery
+        ####################
+        self._queue = []
+        for test in tests:
+            self.log.info("Found test {}.{}".format(test.__module__, test.__qualname__))
+            self._queue.append(test)
+        self.ntests = len(self._queue)
 
-        handle = simulator.get_root_handle(self._root_name)
+        if not self._queue:
+            self.log.warning("No tests were discovered")
 
-        self._dut = cocotb.handle.SimHandle(handle) if handle else None
+        self._queue.sort(key=lambda test: (test.stage, test._id))
 
-        if self._dut is None:
-            raise AttributeError("Can not find Root Handle (%s)" %
-                                 self._root_name)
+        # Process Hooks
+        ###################
+        for hook in hooks:
+            self.log.info("Found hook {}.{}".format(hook.__module__, hook.__qualname__))
+            self._init_hook(hook)
 
-        # Test Discovery
-        ####################
+    @classmethod
+    def from_discovery(cls, dut: SimHandle):
+        """
+        Obtains the test and hook lists by discovery.
+
+        See :envvar:`MODULE` and :envvar:`TESTCASE` for details on how tests are discovered.
+
+        Args:
+            dut (SimHandle): The root handle to pass into test functions.
+        """
+        tests = cls._discover_tests()
+        hooks = cls._discover_hooks()
+        return cls(dut, tests, hooks)
+
+    @staticmethod
+    def _discover_tests() -> Iterable[Test]:
+        """
+        Discovers tests in files automatically.
+
+        See :envvar:`MODULE` and :envvar:`TESTCASE` for details on how tests are discovered.
+        """
+        log = SimLog("cocotb.regression")
```

Comment:
Any reason not to put this at the top of the file at this point, perhaps as `_logger` to keep it private?

Reply:
I think this can be done to a few files so I'll do a sweep as a PR. Is there a better way than `_logger = logging.getLogger("cocotb.{}".format(__NAME__))`? Maybe some way that can detect cocotb as the package and return a package-relative name?

Reply:
`__name__` is already `"cocotb.regression"` here

Reply:
Could you link me to documentation to explain how it gets that name? I can't seem to find anything.

Reply:
Some discussion here: https://stackoverflow.com/q/15883526/102441

Reply:
I added the logger as requested using `__name__`.